File: sdks/python/fileops/files.py
"""
Files Client
"""

from typing import Optional, Dict, Any, BinaryIO
from datetime import datetime
import httpx
from .types import File, Folder, FileVersion, PaginatedResponse


class FilesClient:
    """Files and folders management client."""

    def __init__(self, http: httpx.Client):
        self._http = http

    def list(
        self,
        folder_id: Optional[str] = None,
        page: int = 1,
        limit: int = 20,
        sort: Optional[str] = None,
        order: str = "asc",
    ) -> PaginatedResponse:
        """List files."""
        params = {"page": page, "limit": limit, "order": order}
        if folder_id:
            params["folderId"] = folder_id
        if sort:
            params["sort"] = sort

        response = self._http.get("/files", params=params)
        response.raise_for_status()
        return self._parse_paginated_response(response.json())

    def get(self, file_id: str) -> File:
        """Get file by ID."""
        response = self._http.get(f"/files/{file_id}")
        response.raise_for_status()
        return self._parse_file(response.json())

    def upload(
        self,
        file: BinaryIO,
        filename: str,
        mime_type: Optional[str] = None,
        folder_id: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> File:
        """Upload a file."""
        # Initialize upload
        init_response = self._http.post(
            "/files/upload",
            json={
                "filename": filename,
                "size": 0,  # Will be set by content
                "mimeType": mime_type or "application/octet-stream",
                "folderId": folder_id,
                "metadata": metadata or {},
            },
        )
        init_response.raise_for_status()
        upload_data = init_response.json()

        # Upload content
        files = {"file": (filename, file, mime_type or "application/octet-stream")}
        upload_response = httpx.put(
            upload_data["uploadUrl"],
            content=file.read(),
            headers={"Content-Type": mime_type or "application/octet-stream"},
        )
        upload_response.raise_for_status()

        # Complete upload
        complete_response = self._http.post(
            f"/files/upload/{upload_data['uploadId']}/complete"
        )
        complete_response.raise_for_status()
        return self._parse_file(complete_response.json())

    def download_url(self, file_id: str, version: Optional[int] = None) -> str:
        """Get download URL."""
        params = {}
        if version:
            params["version"] = version

        response = self._http.get(f"/files/{file_id}/download", params=params)
        response.raise_for_status()
        return response.json()["downloadUrl"]

    def update(
        self,
        file_id: str,
        name: Optional[str] = None,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> File:
        """Update file metadata."""
        payload = {}
        if name:
            payload["name"] = name
        if metadata:
            payload["metadata"] = metadata

        response = self._http.patch(f"/files/{file_id}", json=payload)
        response.raise_for_status()
        return self._parse_file(response.json())

    def delete(self, file_id: str) -> None:
        """Delete file."""
        response = self._http.delete(f"/files/{file_id}")
        response.raise_for_status()

    def get_versions(self, file_id: str) -> list:
        """Get file versions."""
        response = self._http.get(f"/files/{file_id}/versions")
        response.raise_for_status()
        return [self._parse_version(v) for v in response.json()]

    # Folder operations
    def list_folders(self, parent_id: Optional[str] = None) -> list:
        """List folders."""
        params = {}
        if parent_id:
            params["parentId"] = parent_id

        response = self._http.get("/folders", params=params)
        response.raise_for_status()
        return [self._parse_folder(f) for f in response.json()]

    def create_folder(self, name: str, parent_id: Optional[str] = None) -> Folder:
        """Create folder."""
        response = self._http.post(
            "/folders", json={"name": name, "parentId": parent_id}
        )
        response.raise_for_status()
        return self._parse_folder(response.json())

    def delete_folder(self, folder_id: str, recursive: bool = False) -> None:
        """Delete folder."""
        response = self._http.delete(
            f"/folders/{folder_id}", params={"recursive": recursive}
        )
        response.raise_for_status()

    def _parse_file(self, data: dict) -> File:
        return File(
            id=data["id"],
            name=data["name"],
            path=data["path"],
            mime_type=data["mimeType"],
            size=data["size"],
            checksum=data["checksum"],
            version=data["version"],
            status=data["status"],
            metadata=data.get("metadata", {}),
            created_by=data["createdBy"],
            created_at=datetime.fromisoformat(data["createdAt"].replace("Z", "+00:00")),
            updated_at=datetime.fromisoformat(data["updatedAt"].replace("Z", "+00:00")),
        )

    def _parse_folder(self, data: dict) -> Folder:
        return Folder(
            id=data["id"],
            name=data["name"],
            path=data["path"],
            parent_id=data.get("parentId"),
            created_by=data["createdBy"],
            created_at=datetime.fromisoformat(data["createdAt"].replace("Z", "+00:00")),
            updated_at=datetime.fromisoformat(data["updatedAt"].replace("Z", "+00:00")),
        )

    def _parse_version(self, data: dict) -> FileVersion:
        return FileVersion(
            version=data["version"],
            size=data["size"],
            checksum=data["checksum"],
            created_at=datetime.fromisoformat(data["createdAt"].replace("Z", "+00:00")),
            created_by=data["createdBy"],
        )

    def _parse_paginated_response(self, data: dict) -> PaginatedResponse:
        return PaginatedResponse(
            data=[self._parse_file(f) for f in data["data"]],
            page=data["pagination"]["page"],
            limit=data["pagination"]["limit"],
            total=data["pagination"]["total"],
            total_pages=data["pagination"]["totalPages"],
            has_next=data["pagination"]["hasNext"],
            has_prev=data["pagination"]["hasPrev"],
        )

File: sdks/typescript/src/files.ts
/**
 * Files Client
 */

import { HttpClient } from './http';
import {
  File,
  Folder,
  FileVersion,
  UploadOptions,
  UploadResult,
  DownloadOptions,
  DownloadResult,
  CreateFolderRequest,
  ShareLink,
  CreateShareOptions,
  PaginationParams,
  PaginatedResponse,
  SearchOptions,
} from './types';

export class FilesClient {
  constructor(private http: HttpClient) {}

  // File operations

  /**
   * List files
   */
  async list(
    params?: PaginationParams & { folderId?: string }
  ): Promise<PaginatedResponse<File>> {
    return this.http.get<PaginatedResponse<File>>('/files', { params });
  }

  /**
   * Get file by ID
   */
  async get(fileId: string): Promise<File> {
    return this.http.get<File>(`/files/${fileId}`);
  }

  /**
   * Initialize file upload
   */
  async initUpload(options: UploadOptions): Promise<UploadResult> {
    const response = await this.http.post<{
      uploadId: string;
      uploadUrl: string;
      chunkSize: number;
      expiresAt: string;
    }>('/files/upload', {
      filename: options.filename,
      size: options.size,
      mimeType: options.mimeType,
      folderId: options.folderId,
      metadata: options.metadata,
    });

    return {
      ...response,
      expiresAt: new Date(response.expiresAt),
    };
  }

  /**
   * Upload a file
   * Note: For browser environments, use FormData. For Node.js, use streams.
   */
  async upload(
    file: Blob | Buffer,
    options: UploadOptions
  ): Promise<File> {
    // Initialize upload
    const { uploadUrl, uploadId } = await this.initUpload(options);

    // For simple uploads, use the direct upload URL
    const formData = new FormData();
    formData.append('file', file as Blob, options.filename);

    // Upload the file
    await fetch(uploadUrl, {
      method: 'PUT',
      body: file,
      headers: {
        'Content-Type': options.mimeType || 'application/octet-stream',
      },
    });

    // Complete the upload
    return this.completeUpload(uploadId);
  }

  /**
   * Complete an upload
   */
  async completeUpload(uploadId: string): Promise<File> {
    return this.http.post<File>(`/files/upload/${uploadId}/complete`);
  }

  /**
   * Abort an upload
   */
  async abortUpload(uploadId: string): Promise<void> {
    await this.http.delete(`/files/upload/${uploadId}`);
  }

  /**
   * Get download URL
   */
  async getDownloadUrl(
    fileId: string,
    options?: DownloadOptions
  ): Promise<DownloadResult> {
    const response = await this.http.get<{
      downloadUrl: string;
      expiresAt: string;
    }>(`/files/${fileId}/download`, {
      params: options,
    });

    return {
      downloadUrl: response.downloadUrl,
      expiresAt: new Date(response.expiresAt),
    };
  }

  /**
   * Download file (returns URL for redirect or fetch)
   */
  async download(
    fileId: string,
    options?: DownloadOptions
  ): Promise<string> {
    const { downloadUrl } = await this.getDownloadUrl(fileId, options);
    return downloadUrl;
  }

  /**
   * Update file metadata
   */
  async update(
    fileId: string,
    data: { name?: string; metadata?: Record<string, unknown> }
  ): Promise<File> {
    return this.http.patch<File>(`/files/${fileId}`, data);
  }

  /**
   * Move file to folder
   */
  async move(fileId: string, folderId: string): Promise<File> {
    return this.http.post<File>(`/files/${fileId}/move`, { folderId });
  }

  /**
   * Copy file
   */
  async copy(
    fileId: string,
    options?: { folderId?: string; name?: string }
  ): Promise<File> {
    return this.http.post<File>(`/files/${fileId}/copy`, options);
  }

  /**
   * Delete file
   */
  async delete(fileId: string): Promise<void> {
    await this.http.delete(`/files/${fileId}`);
  }

  /**
   * Restore deleted file
   */
  async restore(fileId: string): Promise<File> {
    return this.http.post<File>(`/files/${fileId}/restore`);
  }

  // Version operations

  /**
   * Get file versions
   */
  async getVersions(fileId: string): Promise<FileVersion[]> {
    return this.http.get<FileVersion[]>(`/files/${fileId}/versions`);
  }

  /**
   * Restore file to specific version
   */
  async restoreVersion(fileId: string, version: number): Promise<File> {
    return this.http.post<File>(`/files/${fileId}/versions/${version}/restore`);
  }

  // Search

  /**
   * Search files
   */
  async search(options: SearchOptions): Promise<PaginatedResponse<File>> {
    return this.http.get<PaginatedResponse<File>>('/files/search', {
      params: {
        ...options,
        createdAfter: options.createdAfter?.toISOString(),
        createdBefore: options.createdBefore?.toISOString(),
      },
    });
  }

  // Folder operations

  /**
   * List folders
   */
  async listFolders(parentId?: string): Promise<Folder[]> {
    return this.http.get<Folder[]>('/folders', {
      params: { parentId },
    });
  }

  /**
   * Get folder by ID
   */
  async getFolder(folderId: string): Promise<Folder> {
    return this.http.get<Folder>(`/folders/${folderId}`);
  }

  /**
   * Create folder
   */
  async createFolder(data: CreateFolderRequest): Promise<Folder> {
    return this.http.post<Folder>('/folders', data);
  }

  /**
   * Update folder
   */
  async updateFolder(folderId: string, data: { name: string }): Promise<Folder> {
    return this.http.patch<Folder>(`/folders/${folderId}`, data);
  }

  /**
   * Move folder
   */
  async moveFolder(folderId: string, parentId: string): Promise<Folder> {
    return this.http.post<Folder>(`/folders/${folderId}/move`, { parentId });
  }

  /**
   * Delete folder
   */
  async deleteFolder(folderId: string, recursive?: boolean): Promise<void> {
    await this.http.delete(`/folders/${folderId}`, {
      params: { recursive },
    });
  }

  // Sharing operations

  /**
   * Create share link
   */
  async createShare(options: CreateShareOptions): Promise<ShareLink> {
    const response = await this.http.post<{
      id: string;
      fileId: string;
      url: string;
      permissions: string;
      password: boolean;
      expiresAt?: string;
      accessCount: number;
      createdAt: string;
    }>(`/files/${options.fileId}/share`, {
      permissions: options.permissions,
      password: options.password,
      expiresAt: options.expiresAt?.toISOString(),
      maxAccesses: options.maxAccesses,
    });

    return {
      ...response,
      permissions: response.permissions as any,
      expiresAt: response.expiresAt ? new Date(response.expiresAt) : undefined,
      createdAt: new Date(response.createdAt),
    };
  }

  /**
   * List share links for a file
   */
  async listShares(fileId: string): Promise<ShareLink[]> {
    const response = await this.http.get<any[]>(`/files/${fileId}/shares`);
    return response.map((s) => ({
      ...s,
      expiresAt: s.expiresAt ? new Date(s.expiresAt) : undefined,
      createdAt: new Date(s.createdAt),
    }));
  }

  /**
   * Delete share link
   */
  async deleteShare(fileId: string, shareId: string): Promise<void> {
    await this.http.delete(`/files/${fileId}/shares/${shareId}`);
  }

  /**
   * Get file checksum
   */
  async getChecksum(fileId: string): Promise<{ checksum: string; algorithm: string }> {
    return this.http.get(`/files/${fileId}/checksum`);
  }

  /**
   * Verify file integrity
   */
  async verifyIntegrity(fileId: string): Promise<{
    valid: boolean;
    expectedChecksum: string;
    actualChecksum: string;
  }> {
    return this.http.post(`/files/${fileId}/verify`);
  }

  /**
   * Get storage usage
   */
  async getUsage(): Promise<{
    used: number;
    limit: number;
    percentage: number;
  }> {
    return this.http.get('/files/usage');
  }

  /**
   * Get recently accessed files
   */
  async getRecent(limit?: number): Promise<File[]> {
    return this.http.get<File[]>('/files/recent', {
      params: { limit },
    });
  }

  /**
   * Get starred/favorite files
   */
  async getStarred(): Promise<File[]> {
    return this.http.get<File[]>('/files/starred');
  }

  /**
   * Star/favorite a file
   */
  async star(fileId: string): Promise<void> {
    await this.http.post(`/files/${fileId}/star`);
  }

  /**
   * Unstar a file
   */
  async unstar(fileId: string): Promise<void> {
    await this.http.delete(`/files/${fileId}/star`);
  }
}
